common/js/src/messaging/single-message-based-channel.js

```diff
@@ -0,0 +1,141 @@
+goog.require('GoogleSmartCard.MessageChannelPinging.PingResponder');
+goog.require('GoogleSmartCard.MessageChannelPinging.Pinger');
+goog.require('GoogleSmartCard.TypedMessage');
+goog.require('goog.asserts');
+goog.require('goog.log.Logger');
+goog.require('goog.messaging.AbstractChannel');
+
+goog.scope(function() {
+
+/** @const */
+var GSC = GoogleSmartCard;
+
+/** @const */
+var Pinger = GSC.MessageChannelPinging.Pinger;
+/** @const */
+var PingResponder = GSC.MessageChannelPinging.PingResponder;
+
+/**
+ * This class is a wrapper around the Chrome simple one-time requests (see
+ * <https://developer.chrome.com/extensions/messaging#simple>) that transforms
+ * them into Closure-style message channels (see
+ * <http://google.github.io/closure-library/api/interface_goog_messaging_MessageChannel.html>).
+ *
+ * Apart from adapting the one-time communication mechanisms into the methods of
+ * the goog.messaging.AbstractChannel class, this class enables pinging over
+ * this message channel (see the message-channel-pinging.js file for details).
+ * @param {string} extensionId
+ * @param {function()=} opt_onEstablished
+ * @constructor
+ * @extends goog.messaging.AbstractChannel
+ */
+GSC.SingleMessageBasedChannel = function(extensionId, opt_onEstablished) {
+  SingleMessageBasedChannel.base(this, 'constructor');
+
+  /** @const @type {string} */
+  this.extensionId = extensionId;
+
+  /**
+   * @type {!goog.log.Logger}
+   * @const
+   */
+  this.logger = GSC.Logging.getScopedLogger(
+      'SingleMessageBasedChannel<' + extensionId + '>');
+
+  this.registerDefaultService(this.defaultServiceCallback_.bind(this));
+
+  /** @private */
+  this.pingResponder_ = new PingResponder(this, this.logger);
+
+  /** @private */
+  this.pinger_ = new Pinger(this, this.logger, opt_onEstablished);
+
+  this.logger.fine('Initialized successfully');
+};
+
+/** @const */
+var SingleMessageBasedChannel = GSC.SingleMessageBasedChannel;
+
+goog.inherits(SingleMessageBasedChannel, goog.messaging.AbstractChannel);
+
+/** @override */
+SingleMessageBasedChannel.prototype.send = function(serviceName, payload) {
+  GSC.Logging.checkWithLogger(this.logger, goog.isObject(payload));
+  goog.asserts.assertObject(payload);
+
+  var typedMessage = new GSC.TypedMessage(serviceName, payload);
+  var message = typedMessage.makeMessage();
+  this.logger.finest('Posting a message: ' + GSC.DebugDump.debugDump(message));
+
+  if (this.isDisposed()) {
+    GSC.Logging.failWithLogger(
+        this.logger, 'Failed to post message: the channel is already disposed');
+  }
+
+  chrome.runtime.sendMessage(this.extensionId, message);
+};
+
+/** @override */
+SingleMessageBasedChannel.prototype.disposeInternal = function() {
+  this.pinger_.dispose();
+  this.pinger_ = null;
+
+  this.pingResponder_ = null;
+
+  this.logger.fine('Disposed');
+
+  SingleMessageBasedChannel.base(this, 'disposeInternal');
+};
+
+SingleMessageBasedChannel.prototype.deliverMessage = function(message) {
```

Comment:
Please add JSdoc for this method. Some textual definition would be also helpful, as this method is not inherited.
Also better to place this method before disposeInternal, so that all public methods go ahead of private/protected.